CRM-15861: Offline membership renewal doesn't display priceset choices

litespeedmarc opened the pull request:
This is take two of previous request.  I had last night messed up with a git merge after fetch / rebase instead of git push -f

I expect a few more code style nitpicking, will get those in when they come.

Even though I still don't like that a "non-single membership specific action" is triggered by using an action next to a specific membership, the new flow _is_, after functional review from @agh1, @nganivet, and Virginie (can't find her moniker), _much_ better.  It is more inline with current flows, and doesn't clutter the UI with more buttons as my initial kick at the can.

I will update the JIRA with more details.

---
- [CRM-15861: Offline membership renewal doesn't display priceset choices](https://issues.civicrm.org/jira/browse/CRM-15861)

eileenmcnaughton commented:
looks like a php error snuck in https://test.civicrm.org/job/CiviCRM-Core-PR/12124/console

litespeedmarc commented:
Thanks for pointing it out.  Fixed now. I obviously never tested the
scenario where primary financial type is blank (if that is even possible).

On Wed, Oct 12, 2016 at 8:55 AM, Eileen McNaughton <notifications@github.com
> wrote:

> looks like a php error snuck in https://test.civicrm.org/job/
> CiviCRM-Core-PR/12124/console
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/9216#issuecomment-253205438>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADVy4_HzkltS7QppsgdIZbrPyz_h9Zkbks5qzNjEgaJpZM4KUSMn>
> .
>

eileenmcnaughton commented:
There are deeper errors this time - https://test.civicrm.org/job/CiviCRM-Core-PR/12137/testReport/junit/(root)/CRM_Member_Form_MembershipRenewalTest/testSubmit/

It might be just handling the enotice. I would argue price set id probably should always be set so would be reluctant to 'just handle it'

eileenmcnaughton commented:
This looks incredibly valuable! I think it's going to be non-trivial to review and it has missed 4.7.13 but I would dearly like it to go into 4.7.14

The failing tests could be fixed by handling the enotice but we probably want price-set specific tests in the mix too.

I found one issue whereby the price set id is not quite working when renewing a price set one BACK to a non price set one. My preference would be that the front end form sets the price set id to the default one in this case & we handle the flow as similar as possible - with just the form presenting it differently. From a code point of view the default priceset IS a priceset. The only difference really is you can override the financial type from the form later.

Anyway, I'm still looking at this & I hope @seamuslee001 will as well. It would be good to get this rebased over the latest master & perhaps you could set us up to push to your fork to collaborate there?

litespeedmarc commented:
I've been tied down with other work, unable to look at this until my last set of commits a week ago or so.  I've added @seamuslee001 and @eileenmcnaughton as collaborators. I am working in a CRM-15861 branch off my fork.

Can't look at this today. But I do have a "free" Sunday, where I'll get to spend most of my time on this.

eileenmcnaughton commented:
I'm not going to be able to look tomorrow - it will be at least a week now before I can put much more time into it

litespeedmarc commented:
Thanks for the issue on renewing on membership that doesn't have a price set, but previously did.  I've adjusted the SQL.  Should be good now, and will still default the best price set if user does pick to "switch to price set view".

Unit test was easy fix.

I also picked up an (serious) issue with a UDF call (page would just crash, with certain custom fields defined).

I agree as well on unit tests (incredibly so). I still need sometime to get my environment going with running unit tests.  If it is acceptable, I would be very happy to take that on as a separate CRM.  I say this because the client I'm working for, really would like this change, and the sooner we can get it to them the better.

eileenmcnaughton commented:
In terms of timing - we are looking at trying to get this into the release that will go out on the first Wed of Dec. 

I'm going to try to find some more time on this next week

litespeedmarc commented:
Rebased & Squashed.

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2508,4 +2516,93 @@ public static function getContactsCancelledMembership($contactID, $isTest = FALS
+            li.price_field_id,
+            li.price_field_value_id,
+            co.receive_date
+          from (select distinct member_of_contact_id, con.display_name from civicrm_membership_type left join civicrm_contact con on con.id = member_of_contact_id) org
```
Just a stylistic not - we normally use the mysql convention of uppercasing FROM, INNER JOIN, WHERE, ORDER BY

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2508,4 +2516,93 @@ public static function getContactsCancelledMembership($contactID, $isTest = FALS
+   *   Array (int, int, int): The membership orgs that contact is part of.
+   */
+  public static function getActiveContactMemberships($contact_id) {
+    $sql = "select distinct member_of_contact_id
```
ditto SELECT DISTINCT

eileenmcnaughton commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
We should get rid of this empty line

eileenmcnaughton commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -68,8 +69,7 @@ public static function create(&$params) {
     $validatePriceSet = TRUE;
     if (!empty($params['extends']) && is_array($params['extends'])) {
-      if (!array_key_exists(CRM_Core_Component::getComponentID('CiviEvent'), $params['extends'])
-        || !array_key_exists(CRM_Core_Component::getComponentID('CiviMember'), $params['extends'])
```
just re-formatting? why?

eileenmcnaughton commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -786,7 +784,7 @@ public static function processAmount($fields, &$params, &$lineItem, $component =
     foreach ($fields as $id => $field) {
       if (empty($params["price_{$id}"]) ||
-        (empty($params["price_{$id}"]) && $params["price_{$id}"] == NULL)
+          (empty($params["price_{$id}"]) && $params["price_{$id}"] == NULL)
```
why this?

eileenmcnaughton commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -1034,9 +1032,9 @@ public static function buildPriceSet(&$form) {
-      'CRM_Contribute_Form_Contribution',
-      'CRM_Member_Form_Membership',
-    ))) {
+          'CRM_Contribute_Form_Contribution',
```
hmm - some more odd re-formatting

litespeedmarc commented:
Not sure what happened there.  I will, on occasion, reformat stuff if it
helps readability, but can't say that's the case here.  My guess, is that I
had some odd Netbeans auto-format on save setting (I'm using PHPStorm now).

As far as I remember (add see from just comparing now), the only diffs were
additions.

Reinstating original formatting.

On Sun, Oct 30, 2016 at 11:38 PM, Eileen McNaughton <
notifications@github.com> wrote:

> *@eileenmcnaughton* commented on this pull request.
> ------------------------------
>
> In CRM/Price/BAO/PriceSet.php
> <https://github.com/civicrm/civicrm-core/pull/9216#pullrequestreview-6383474>
> :
>
> > @@ -1373,6 +1358,27 @@ public static function checkPermission($sid) {
>    }
>
>    /**
> +   *
> +   * Return the number of active price sets that include memberships.
> +   *
> +   * @return int
> +   *    e.g., 3 to indicate that there are 3 active price sets that
> +   *          use memberships
> +   */
> +  public static function getCountOfPriceSetsWithMemberships() {
>
> I feel like this addition is the only real change in this file?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/9216#pullrequestreview-6383474>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADVy45GNrY_haA5hPLNJvrt-1Dn6UI8nks5q5WK3gaJpZM4KUSMn>
> .
>

eileenmcnaughton commented:
I'm working through the membership & membership renewal forms and pushing up some minor changes (consolidations of code between them) as I make sense of them

eileenmcnaughton commented:
NB a form where we split out the over-loaded form rendering to a separate form is CRM_Financial_Form_Payment

- the other forms that render those fields call the same functions - ie. CRM_Core_Payment_ProcessorForm::preProcess($this);

eileenmcnaughton commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -445,7 +498,18 @@ public static function formRule($params) {
     }
-    return empty($errors) ? TRUE : $errors;
+
+    if (empty($errors) && !empty($params['price_set_id'])) {
```
I'm not liking this - the 2 classes have a shared parent so let's just call a shared function on the parent

eileenmcnaughton commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -497,20 +561,24 @@ protected function submit() {
+
+    if ($this->_priceSet) {
+      // renewing using a price set, set the default financial type (if not set) to the price set's default financial type.
+      if (empty($this->_params['financial_type_id'])) {
```
We don't need the empty check - one of the few differences between $quickConfig and normal price sets is that for quick config it is possible to override the financial_type_id

litespeedmarc replied:
Finally getting back to this.  So you're saying for normal price sets, you shouldn't be able to override the `financial_type_id`?  Doesn't the UI allow it in a normal price set membership addition (and also in the renew one)?

eileenmcnaughton commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -408,32 +468,75 @@ public function buildQuickForm() {
+    // we pass it along to smarty template, so that it can exit early.
+    $this->assign("priceSetOnly", $this->_priceSetOnly);
+
+    $this->assign('price_set_has_new_orgs', $this->getPriceSetHasNewOrgs());
```
I believe this is now redundant as you used the renewal date as the join date. The function and the js that hands of this var should be done for too now

eileenmcnaughton commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -1747,6 +1767,205 @@ public static function getNonDeductibleAmountFromPriceSet($priceSetId, $lineItem
+    $member_of_contact_id = CRM_Member_BAO_MembershipType::getFieldValue('CRM_Member_BAO_MembershipType', $memberhsip_type_id, 'member_of_contact_id');
+    $contact_id = CRM_Member_BAO_Membership::getFieldValue('CRM_Member_BAO_Membership', $membership_id, 'contact_id');
+
+    // we've got an array, above, of all orgs mapped on price set id.
```
I have a problem with this logic. I think that since we are starting from a particular membership then the 'best' option is one that includes that membership type id. I am not sure that there is much value in doing more than trying to find the first match against that id - since the idea people want the one with the most membership matches is really just guess work.

Note that if this reference to getActiveContactMemberships goes then that function is fully redundant

litespeedmarc replied:
The client I'm working with has several price sets to pick from, each price set covering several orgs & types, and with overlap in various membership orgs & types between the price sets.  Typically, one price set handles all of a member's memberships.  Without going for the "most", there is an extra owness (sp?) on the user to "figure out" which price set would "work" the best.  It'd be workable without it, for sure, but not fully optimized from a user (or at least this client's) perspective.

It could alternatively be handled with a hook, but I figure 1) it would be more work through a hook, 2) although somewhat cumbersome SQL, going with the idea that if you switch to "price set", you likely do want to cover as many of the memberships as possible -- generally speaking, and 3) if it's good for this client, it would be likely be of value to others.

eileenmcnaughton commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -716,4 +742,101 @@ protected function submit() {
+    $this->_params = $this->setPriceSetParameters($this->_params);
+    CRM_Price_BAO_PriceSet::processAmount($this->_priceSet['fields'], $this->_params, $lineItem[$this->_priceSetId], NULL, $this->_priceSetId);
+
+    // litespeedmarc: commented out, this only made sense for price set based renewals
```
In testing the financial type is also being skipped where it has been specified and the membership type financial type is winning out at the contribution and line item level

JoeMurray replied:
@eileenmcnaughton I believe that is the correct behaviour. @litespeedmarc you don't need to respond to this comment.

eileenmcnaughton commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -716,4 +742,101 @@ protected function submit() {
+  }
+
+
+  private function submitRenewMembershipSingle($membershiId, $membershpiTypeId, $contributionRecurID) {
```
parameter name misspelling

eileenmcnaughton commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -1373,6 +1358,27 @@ public static function checkPermission($sid) {
+   *    e.g., 3 to indicate that there are 3 active price sets that
+   *          use memberships
+   */
+  public static function getCountOfPriceSetsWithMemberships() {
```
I don't think this function is being called

JoeMurray replied:
There is no call of this function, so please remove from PR.

litespeedmarc replied:
Removed in next commit.

eileenmcnaughton commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -1747,6 +1765,205 @@ public static function getNonDeductibleAmountFromPriceSet($priceSetId, $lineItem
+
+  public static function setPriceSetDefaultsToLastUsedValues(&$defaults, $priceSet, $contactID) {
+
+    // this is detailed info on how membership orgs, and the best membership type & price set field to use
```
I personally think if a comment has to made right before a function is called it suggests the comment belongs in the comment block for the function

JoeMurray replied:
Please move create a DocBlock for the function and move the comment into it as requested? Thx!

litespeedmarc replied:
The comment was more specific to the first line.  I missed the doc block part.  I've got that coming now.

eileenmcnaughton commented on templates/CRM/Member/Form/MembershipPriceSetCommon.tpl:
```diff
@@ -0,0 +1,10 @@
+{if $priceSetId}
+  {include file="CRM/Price/Form/PriceSet.tpl" context="standalone" extends="Membership"}
+  <script>
+    {if $price_set_has_new_orgs}
```
actually this join date argy bargy can go now too

JoeMurray replied:
I don't quite follow what @eileenmcnaughton means here. My skim on what is happening is that if the price set includes one or more organizations that the contact does not have a membership in (or current membership? or current membership that has not expired?) then you want to show a Join date description field. But I think the description is more confusing than enlightening, so I'd also recommend getting rid of this join date description show/hide stuff.

JoeMurray replied:
https://github.com/civicrm/civicrm-core/pull/9216/files#diff-47b4782462c2158a1f4ee86a5145dc0cR4

litespeedmarc replied:
Either or works for me.  Removed in my next commit.

eileenmcnaughton commented on templates/CRM/Member/Form/MembershipRenewal.tpl:
```diff
@@ -23,6 +23,17 @@
+{include file="CRM/Member/Form/MembershipPriceSetCommon.tpl"}
+{if $priceSetId}
+  <script>
+    {if $price_set_has_new_orgs}
```
opps I duplicated it and it can still go

eileenmcnaughton commented on templates/CRM/Member/Form/MembershipRenewal.tpl:
```diff
@@ -303,5 +402,37 @@
       }
     }
+
+    function buildAmount( ) {
```
I'd feel much better if this was consolidated into a function shared with the membership.tpl form

JoeMurray replied:
Again, please respond either by adopting suggestion or explaining why it is not done/should not be done.

litespeedmarc replied:
Same comment here.  The two have similarities, but are still significantly different.

eileenmcnaughton commented on templates/CRM/Member/Form/MembershipRenewal.tpl:
```diff
@@ -226,8 +271,62 @@
+
+      function showPriceSet() {
+      {/literal}
+        {if $price_set_has_new_orgs}{literal}
```
This var is not gone so the show_hide can too. Perhaps with it gone we can share the function with membership.tpl

JoeMurray replied:
Since you are not defining this $price_set_has_new_orgs variable in the PR, it will never be set, so let's remove this code. With it gone, @eileenmcnaughton is suggesting the function showPriceSet can be moved to a different template file. Please do that or explain why you don't think it is necessary. In general, when you get detailed comments that someone has put time and effort into creating, it's a good idea to respond to each individually, and not expect that an overall comment is sufficient.

litespeedmarc replied:
`showPriceSet` still remains fairly specific to the membership renewal.  We *might* be able to generalize it, but I do think this might border coincidental abstraction and in the long term increase to brittleness.

litespeedmarc commented:
This time zone <> thing does wonders!  I wake up, and there's all sorts of
new things!

This, all looked fine.  In theory :)

I always cringe at expressions like: $this->allMembershipTypeDetails[$this->
_memType]['member_of_contact_id']
Demeter violation!  Though that's a hard one to stick to when working with
array laden Php & CiviCRM.  A local variable assignment can help clarify
things.  Maybe $this->assignMembershipOrgDetails()?

I will need tonight to get these changes tested locally, and if all's good
will pull the PR onto the client's test system (they were starting testing
of this functionality today, will hold them off 'till tomorrow).  They are
a good environment to test with: heavy price set uses, multiple membership
type orgs, with a custom extension that does further business specific
optimizations.

I'd expect to hear back some more on Thursday evening or so -- Friday
morning for you :)

Thanks for all the work!


On Mon, Oct 31, 2016 at 7:55 AM, Eileen McNaughton <notifications@github.com
> wrote:

> *@eileenmcnaughton* commented on this pull request.
> ------------------------------
>
> In CRM/Member/Form/MembershipRenewal.php
> <https://github.com/civicrm/civicrm-core/pull/9216#pullrequestreview-6419103>
> :
>
> > @@ -445,7 +498,18 @@ public static function formRule($params) {
>          $errors['payment_instrument_id'] = ts('Payment Method is a required field.');
>        }
>      }
> -    return empty($errors) ? TRUE : $errors;
> +
> +    if (empty($errors) && !empty($params['price_set_id'])) {
>
> I'm not liking this - the 2 classes have a shared parent so let's just
> call a shared function on the parent
> ------------------------------
>
> In CRM/Member/Form/MembershipRenewal.php
> <https://github.com/civicrm/civicrm-core/pull/9216#pullrequestreview-6419103>
> :
>
> > @@ -497,20 +561,24 @@ protected function submit() {
>      $this->assign('receive_date', $this->_params['receive_date']);
>      $this->processBillingAddress();
>      list($userName) = CRM_Contact_BAO_Contact_Location::getEmailDetails(CRM_Core_Session::singleton()->get('userID'));
> -    $this->_params['total_amount'] = CRM_Utils_Array::value('total_amount', $this->_params,
> -      CRM_Core_DAO::getFieldValue('CRM_Member_DAO_MembershipType', $this->_memType, 'minimum_fee')
> -    );
> -    $this->_membershipId = $this->_id;
> -    $customFieldsFormatted = CRM_Core_BAO_CustomField::postProcess($this->_params,
> -      $this->_id,
> -      'Membership'
> -    );
> -    if (empty($this->_params['financial_type_id'])) {
> -      $this->_params['financial_type_id'] = CRM_Core_DAO::getFieldValue('CRM_Member_DAO_MembershipType', $this->_memType, 'financial_type_id');
> +    $this->assign('contactID', $this->_contactID);
> +
> +    if ($this->_priceSet) {
> +      // renewing using a price set, set the default financial type (if not set) to the price set's default financial type.
> +      if (empty($this->_params['financial_type_id'])) {
>
> We don't need the empty check - one of the few differences between
> $quickConfig and normal price sets is that for quick config it is possible
> to override the financial_type_id
> ------------------------------
>
> In CRM/Member/Form/MembershipRenewal.php
> <https://github.com/civicrm/civicrm-core/pull/9216#pullrequestreview-6419103>
> :
>
> > +          NULL, array('onchange' => "buildAmount( this.value );")
> +    );
> +
> +    // pass the price set id of default (or selected) price set back to smarty
> +    $this->assign("priceSetId", $this->_priceSetId);
> +    // set default on price set to value determined in buildQuickForm
> +    $this->set("priceSetId", $this->_priceSetId);
> +    // add all elements that are part of price set (these will be hidden to start)
> +    // but will correspond to "last price set used".
> +    CRM_Price_BAO_PriceSet::buildPriceSet($this);
> +
> +    // _priceSetOnly is set as request param via ajax onChange of price set field
> +    // we pass it along to smarty template, so that it can exit early.
> +    $this->assign("priceSetOnly", $this->_priceSetOnly);
> +
> +    $this->assign('price_set_has_new_orgs', $this->getPriceSetHasNewOrgs());
>
> I believe this is now redundant as you used the renewal date as the join
> date. The function and the js that hands of this var should be done for too
> now
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/9216#pullrequestreview-6419103>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADVy436SHmJ4dFW2kyPpssIRL3W1xtDqks5q5dcngaJpZM4KUSMn>
> .
>

eileenmcnaughton commented:
I've been wrestling with this and I still can't quite get to grips with it. It feels like we are introducing an element of randomness. If I can buy 2 memberships at once which one is renewing the existing membership? Does the other one create a new membership? What if another current membership exists? Of the same type? Or a different type? What if it is expired.

Realistically the UI should allow these things to be specified. It feels like a candidate for leap by extension - except so much time has already gone into it, but I worry we are going to be hammering out these nuances forever and a day if we don't get a UI that really does permit us to specify what we actually want to happen.

What was the usecase you were originally tackling? I have had requests for this sort of thing before - but it was always for the membership renewal + contribution scenario rather than multiple memberships

litespeedmarc commented:
Randomness: Would be the same today as renewing using a price set, through
the on-line membership contribution page.

If you buy 2 memberships at once: Memberships are renewed based on
membership org (this is how it happens on the on-line contribution page).
One can certainly poke wholes at the existing on-line logic (easily
enough): dates, expired, differing existing start/end dates.  However,
using an 80/20 engineering approach, the current logic has worked well, and
can easily extend to off-line renewals.  The conundrum is really, does the
triggering of the price set based renewals really belong where they are
at?  After all, you could technically end-up not even renewing the existing
membership, and only creating two new ones.  So the answer is *NO* (it
doesn't belong).  My first "kick-at-the-can" had it as a general action,
available on the membership tab (i.e., "renew memberships using price
set"), and then as a button next to "create new membership".  But end-user
reviews (Véronique @ CiviHelpDesk, Andrew @ AGS, Alicia @ AGS, and even
with this client), showed that users naturally went to what they already
new: "renew".

I'm not sure what you mean by "candidate for leap by extension".  Handle it
via extension instead (and then integrate)?  Possibly -- though I think it
might be more work, and more brittle.  I might have gone down that path had
there not already been a core use case where it was being used (on-line).

We could most definitively be spending a lot of time hammering nuances.  I
find myself doing this as well!  In the end, I wanted to be pragmatic, keep
in-line with on-line renewals, get something done over nothing, and
(believing that this belongs in core), not put the client in a situation
where I'd be mavericking their core code.

The use case I'm tackling can be summed up with this price set:
https://issues.civicrm.org/jira/secure/attachment/53976/53976_SamplePriceSet.pdf.
One price set handling, up to 7 societes!  In the implementation chosen for
this client, subscriptions are also handled as societies.

On Tue, Nov 1, 2016 at 1:29 AM, Eileen McNaughton <notifications@github.com>
wrote:

> I've been wrestling with this and I still can't quite get to grips with
> it. It feels like we are introducing an element of randomness. If I can buy
> 2 memberships at once which one is renewing the existing membership? Does
> the other one create a new membership? What if another current membership
> exists? Of the same type? Or a different type? What if it is expired.
>
> Realistically the UI should allow these things to be specified. It feels
> like a candidate for leap by extension - except so much time has already
> gone into it, but I worry we are going to be hammering out these nuances
> forever and a day if we don't get a UI that really does permit us to
> specify what we actually want to happen.
>
> What was the usecase you were originally tackling? I have had requests for
> this sort of thing before - but it was always for the membership renewal +
> contribution scenario rather than multiple memberships
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/9216#issuecomment-257491937>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADVy42a3rIOWvYHQnU2rcr_OY8XqYqt-ks5q5s4_gaJpZM4KUSMn>
> .
>

eileenmcnaughton commented:
The big difference here compared to the online version is that if you go to renew a membership and then choose a membership type from a different organisation it then extends and changes the existing one. I find the whole upsell problematic. In order to prevent upselling it is common for sites to configure an organisation for each membership type.  Here however, they could accidentally  upselling to a membership type in a different org, and it's not clear which memberships would be chosen as new & which would be chosen as renewals. If the existing membership were only extended in the case of it being the same membership type, or if we filtered out memberships of a different organisation I'd feel less nervous. 

(as an aside I have tended to deal with it by using the online form now that you can do that on behalf of a different contact)

litespeedmarc commented:
I started testing locally, and noticed an issue right away.  If I get an edit error using a price set, after I "fix" the edit error,  it ignores my price set choices and randomly assigns a membership!  Video (get some popcorn), showing pre & post.  http://www.scibrazeau.com/CRM-15861/CRM15861Vid1.mov

I looked quickly, looks related maybe to Form/Membership.buildQuickForm $getOnlyPriceSetElements being set to TRUE after an edit?  Not sure which change made that happen (nothing jumped).  Can't look today unfortunately.

litespeedmarc commented:
I missed your last reply before my last, but in my testing of normal on-line price sets, if you don't pick a membership option for an org for which you already have a membership, it simply doesn't touch that member's membership record for that org.  Double checked this against 4.7.12, and that is correct.  This makes perfect sense to me (how it works), even more so on the online, because renewal is driven from the MEMBER, not a membership.

For the aside, I don't see an out-of-the-box way to use on-line price sets for a back office workflow (typically triggered from the member's contact record).  I can pick "do that on behalf of a different contact", but then I'm in a *new* membership workflow and/or have to start looking at de-dup rules & additional contribution pages that don't include member profile fields (unless, I'm missing something, which is very possible).

litespeedmarc commented:
@eileenmcnaughton I fixed the issue that I had pointed out.  Also noticed that I had left a hard coded ID in my SQL!!!  Fixed that too.

Back to you.

Do I need to do anything to approve your changes?  I was able to pull them fine, so my guess is it will all flow, once we get to merge this (hopefully soon :) )

If you think it's good now, will also pull @alifrumin back in, as she was volunteering for some QA as well :)

eileenmcnaughton commented:
Hey - I know I need to look again - I'm definitely not waiting on anything from you at this stage - just trying to find some time

litespeedmarc commented:
Probably most pressing for me, would be to settle our discussions on
functionality. I'm hoping I've convinced you it is good as it is. I've got
two people lined up for PR testing, and then customer will also do
integration testing of the patch (back ported) against 4.7.13.

On Sun, Nov 6, 2016 at 2:28 AM, Eileen McNaughton <notifications@github.com>
wrote:

> Hey - I know I need to look again - I'm definitely not waiting on anything
> from you at this stage - just trying to find some time
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/9216#issuecomment-258665108>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADVy45O0JyUf1kZ-l1WV23tc_zX4dDrqks5q7YGpgaJpZM4KUSMn>
> .
>

litespeedmarc commented:
@eileenmcnaughton , any objections to @alifrumin & some of the CIC staff doing some manual testing using the last set of commits (using official Jenkins build above)?  

If not, feel like giving it a kick at the can @alifrumin ?

eileenmcnaughton commented:
Go ahead!

eileenmcnaughton commented on CRM/Member/Form.php:
```diff
@@ -77,6 +77,23 @@ class CRM_Member_Form extends CRM_Contribute_Form_AbstractEditPayment {
+   * Membership/MembershipRenewal Form used to enter memberships.  When user selects
+   * a PriceSet, we use JavaScript to inject HTML returned by a recursive call to this form.
+   *
+   * This flag controls this recursive call, short-circuiting it to only retrieve the price
```
This is a pattern we have been aggressively removing elsewhere - we need to think about another way for this

litespeedmarc replied:
I was about to comment on this, and then saw your WIP#9387, which is pretty well where my comment was going to head with some extra bits about Angular.

Might be OK if time allows it.  Unfortunately, I'm out of coding cycles, so this could send it for a spin depending on how far you get into this.  Maybe as a separate JIRA?

The words "real price set" also made me think would be nice to have a test case that includes a "real price set" -- the complex one that the CIC uses.

litespeedmarc replied:
That is fair enough.  My main concern would be that it never gets merged, and that this client be "stuck" with a PR that they have to keep rebasing & pulling locally on every upgrade.  But that doesn't sound like it is the case.

Although I'm out of dev cycles for the next month or so, I will certainly provide in the short term any "bug" support on this.  In the new year, you can count on me for #1 unit tests and your PR #9387 (extracting the price set into its own form).

litespeedmarc replied:
@alifrumin, @eileenmcnaughton has talked me into holding off the PR until at the very least unit tests can be done.  I think, ideally testing would cover a more finished set of changes.  So sounds like you'll be holding off a little while still.

Until then, I will manually pull the PR on every customer upgrade (yuk).

alifrumin commented:
@litespeedmarc, let me know when you guys are done with the changes and I can test it.

litespeedmarc commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -493,20 +557,24 @@ protected function submit() {
     $this->assign('receive_date', $this->_params['receive_date']);
     $this->processBillingAddress();
     list($userName) = CRM_Contact_BAO_Contact_Location::getEmailDetails(CRM_Core_Session::singleton()->get('userID'));
-    $this->_params['total_amount'] = CRM_Utils_Array::value('total_amount', $this->_params,
```
Only calculate `total_amount`, using `minimum_fee` for non-price sets renewals.  For price sets ones, this will be calculated.

litespeedmarc commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -493,20 +557,24 @@ protected function submit() {
-      CRM_Core_DAO::getFieldValue('CRM_Member_DAO_MembershipType', $this->_memType, 'minimum_fee')
-    );
-    $this->_membershipId = $this->_id;
-    $customFieldsFormatted = CRM_Core_BAO_CustomField::postProcess($this->_params,
```
Only calculate total_amount, using minimum_fee for non-price sets renewals. For price sets ones, this will be calculated.
_membershipId will only apply for non-price set ones, as their may be multiple memberships being renewed when using price sets.

litespeedmarc replied:
postProcess moved to submitRenewMembershipSingle

litespeedmarc commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -493,20 +557,24 @@ protected function submit() {
+
     $contributionRecurID = NULL;
     $this->assign('membershipID', $this->_id);
-    $this->assign('contactID', $this->_contactID);
```
Moved up

litespeedmarc commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -557,30 +626,28 @@ protected function submit() {
-    if (is_numeric(CRM_Utils_Array::value('num_terms', $this->_params))) {
-      $numRenewTerms = $this->_params['num_terms'];
-    }
-
```
Moved into `submitRenewMembershipSingle`

litespeedmarc commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -593,33 +660,12 @@ protected function submit() {
-        $this->_params['lineItems'] = $lineItem;
-        $this->_params['processPriceSet'] = TRUE;
-      }
-
```
Moved to `submitAddContributionLineItemsToParams`

litespeedmarc replied:
Don't think above comment from me was right. Looking at what was going on here (price set stuff on renewal!?) don't think any of it ever applied.

litespeedmarc commented:
@eileenmcnaughton & @alifrumin .  January has come!  So has Feb, Mar, Apr & Half of May!  

I was only able to pick this up again a bit in mid/late April, and again this week-end.  As per your last comment, work has gone into unit (or rather integration) tests, and additional efforts in rebasing (usual conflicts in overloaded membership classes).

Assuming the latest build will be green (looks like it will as Jenkins is passed the core sections in unit tests), this is ready for a final round of testing, review & [hopefully] a pull into master.

litespeedmarc commented:
Jenkins, retest this please.

litespeedmarc commented:
@eileenmcnaughton , @mlutfy , @alifrumin .  

Pinging again, to see if we could get thoughts on what we can do to get this PR approved into core.

mlutfy commented:
jenkins, test this please

litespeedmarc commented:
Bump

JoeMurray commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -166,51 +162,81 @@ public function preProcess() {
+   *
+   * @return bool
+   */
+  private function getPriceSetHasNewOrgs() {
```
The PR and its commits are no longer properly linked. Here's link to this file in current PR: https://github.com/civicrm/civicrm-core/pull/9216/files#diff-fa0276456bc05f4629f775e6846f62ed

litespeedmarc replied:
I not sure what we mean by switch to renewal_date, or if that comment applies anymore.  Ok, looking more closely, removing the "confusing" description does make it uncalled from what I can tell.  Removed in next commit.

JoeMurray commented on CRM/Member/Form.php:
```diff
@@ -78,6 +77,23 @@ class CRM_Member_Form extends CRM_Contribute_Form_AbstractEditPayment {
   public $_priceSetId;
```
@litespeedmarc it doesn't appear that you have addressed @eileenmcnaughton 's concern here. Could you comment?

litespeedmarc replied:
My apologies for missing the replies here, something happened with my e-mail filters.  @JoeMurray / @eileenmcnaughton  the history is getting a bit unwieldy on this.  I'm not sure which comment we are referring to.

To make things better (or maybe worse) I've just rebased / pushed --force onto this branch, so I'm not sure if that might have affected line numbers.  Is it worth moving into a fresh PR?

JoeMurray commented on templates/CRM/Member/Form/MembershipRenewal.tpl:
```diff
@@ -88,7 +121,12 @@
+        <td>{include file="CRM/common/jcalendar.tpl" elementName=renewal_date}
+        {if $priceSetId}
+        <br />
+        <span class="description hiddenElement" id="join_date_desc">{ts}If you choose a membership type for an organization for which this contact is not member of, then this date will also be used as the 'Member Since' date{/ts}</span>
```
If you really want this description, then how about instead of 'If you choose a membership type for an organization for which this contact is not member of, then this date will also be used as the 'Member Since' date' 
we use:
"If you choose a membership type for an organization in which this contact is not member, this date will also be used as the 'Member Since' date for membership in that organization.'

Personally I think this is superfluous and potentially confusing on-screen help; it merely restates the normal use of a field as though it were somehow being treated specially. I would remove the description and the associated show/hide toggling.

litespeedmarc replied:
Removed.

JoeMurray commented:
Hi @litespeedmarc and @mlutfy - this looks like a lot of work has gone into it, and many but not all points raised by @eileenmcnaughton previously have been addressed. If you could address the remaining ones that I've highlighted, then I'll have @monishdeb look over your next version and do some testing.

eileenmcnaughton commented:
I still feel nervous about this :-( I'm going to be moving dedupe to an extension & this may forge a path for this to go that way but I'm also aware that not only has a tonne of work gone into this but it has been discussed at sprints with @nganivet & @agh1 staff encouraging it as a core improvement. I'd really like to see CiviDesk & AGH involved as well in either getting it over the line (with adequate testing / functional approval) or into an extension with an acceptable maintenance path (e.g ships with core)

nganivet commented:
@eileenmcnaughton Please keep me updated on moving dedupe to an extension, you know I have already committed to moving geolocalization to a 'core' extension once this is in place.

mlutfy commented on CRM/Price/BAO/PriceSet.php:
```diff
@@ -1672,6 +1669,236 @@ public static function getNonDeductibleAmountFromPriceSet($priceSetId, $lineItem
+      $activateByDefault = FALSE;
+    }
+
+    // nothing found.  we want to pick, by default the price set that will cover
```
@litespeedmarc I'm not very comfortable with this, since it can return a different default depending on the time of year, or evolution of membership.

litespeedmarc replied:
The business goal was to always default the best possible default.  So the requirement that I set was this.
1) Use last price set used for membership being renewed (this one seems evident).
2) If that doesn't work, pick a price set that will match all of this contact's memberships

In the case of my client, #2 worked well because of how their data was converted: initial memberships were imported without contributions, so it would be impossible to determine a priceset for those initial conversions.

Objectively, not sure many people would care much about this second rule, specially since it only comes into play after user clicks on "click here" from [this portion](https://issues.civicrm.org/jira/secure/attachment/53277/53277_Screen+Shot+2.png) of the renew membership dialog.

That being said, this would not return a different default depending on time of year/membership evolution (barring, price set config changes and/or deletions/additions of new membership to that contact's actual memberships -- but then all bets are off).

It would be possible, however, to get a different default depending on contact selected, potentially of concern, but also explainable with above rule. Weighing pros (a default in an edge case) vs cons (added complexity + potentially some explainable confusion on default), I'm almost talking myself out of it.

So I'm fine removing this altogether.  Would just need to explain to client (which they'll probably be fine with) and/or add an extension hook spot here.

mlutfy commented:
jenkins, test this please

mlutfy commented:
I have run into a bug:

* Go to Membership > New membership

Only pricesets are displayed. For sites not using membership pricesets, it's not possible to select an org + membership type.

Expected:

![capture d ecran de 2017-08-30 18-18-47](https://user-images.githubusercontent.com/254741/29897591-b37939f0-8daf-11e7-9368-53ed9daaaf11.png)

With the patch:

![capture d ecran de 2017-08-30 18-19-32](https://user-images.githubusercontent.com/254741/29897618-d23b6458-8daf-11e7-953b-6053e3f06486.png)

Seems to be:

```
    // use to build form during form rule.  If request was to "_priceSetOnly"
    // then we're obviously not in a rule, so we don't need to build price set,
    // we we pass FALSE to 'buildPriceSet' argument
    $this->assign('buildPriceSet', !$this->_priceSetOnly);
```

This gets evaluated to TRUE, but the Membership.tpl has:

```
    // load form during form rule.
    {/literal}{if $buildPriceSet}{literal}
    cj( "#totalAmountORPriceSet" ).hide( );
    cj( "#mem_type_id" ).hide( );
    cj('#total_amount').attr("readonly", true);
    cj( "#num_terms_row" ).hide( );
    cj(".crm-membership-form-block-financial_type_id-mode").hide();
    {/literal}{/if}{literal}
```

So if showing pricesets, it hides the `mem_type_id`.

litespeedmarc commented:
Thanks for identifying this issue @mlutfy . This is a lot more of an issue that just not having price sets.  This just wouldn't work "all the time" even with price sets, in the initial membership creation.  Happy in a way that this isn't a subtle error.  It was just a silly mistake.  Never came up with my client because we had an extension that hid the org option and that always defaulted a price set or forced user to pick one.

litespeedmarc commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -570,69 +637,37 @@ protected function submit() {
-    // a function this form shared with other forms.
-    $membershipSource = NULL;
-    if (!empty($this->_params['membership_source'])) {
-      $membershipSource = $this->_params['membership_source'];
```
Moved into submitRenewMembershipSingle

litespeedmarc commented on CRM/Member/Form/MembershipRenewal.php:
```diff
@@ -570,69 +637,37 @@ protected function submit() {
-      $isPending,
-      $contributionRecurID, $membershipSource, $this->_params['is_pay_later'], CRM_Utils_Array::value('campaign_id',
-      $this->_params)
-    );
```
Moved to submitAddContributionLineItemsToParams

mattwire commented:
I seem to be having an issue with the wrong membership being renewed and it's membership type being changed when saving - when I use the "renew" option against an individual membership.

mattwire commented:
@litespeedmarc: See https://github.com/mattwire/civicrm-core/commit/300b313f0c29262b376da7db16eafa877496b07c for my fix for the wrong membership being renewed.

mattwire commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2523,4 +2531,93 @@ public static function getContactsCancelledMembership($contactID, $isTest = FALS
+   *   )
+   *
+   */
+  public static function getContactMemberhipsByMembeshipOrg($contactId) {
```
membeship -> membership

mattwire commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2523,4 +2531,93 @@ public static function getContactsCancelledMembership($contactID, $isTest = FALS
 
+  /**
+   *
+   * Given specified contact, return a map of membeship orgs (that contact
```
membeship -> membership

litespeedmarc commented:
@mattwire .  Thank you so much for spending some time on this.  I keep fearing it will never see day.  Your issue is concerning to me.  However I couldn't replicate it.

As I commented in https://github.com/mattwire/civicrm-core/commit/300b313f0c29262b376da7db16eafa877496b07c, I believe your patch works for you because it is effectively circumventing the payload of CRM-15861.  I'm hoping maybe you can shed a bit more light on replication scenario.

litespeedmarc commented:
In investigating the issue reported by @mattwire, I did notice a couple of minor items related to the refactor to centralize the build price set logic.  Under renewals "choose price set" may not be a viable option in certain scenarios (where one is defaulted), nor is adding an option of "manually specify membership type" as this flow is achieved through the link on the right of the combo box.  Nothing anyone noticed / complained about, but I fixed because I tend to get fussy on these things.

Also fixed typos identified by @mattwire .  It would seem I am incapable of typing membehip.

monishdeb commented:
Closing in favor of https://github.com/civicrm/civicrm-core/pull/11167
